### PR TITLE
Jonmv/async document v1

### PR DESCRIPTION
**Pull request by jonmv**
@bratseth or @bjorncs please review.

### CODE REVIEW

**Comment by bratseth** [comment]
Just to mention it: I think we probably want to keep not_found a success for remove but change the other one.

Regarding providing this or sessions: This API might make people create a lot of sessions, and I think the sessions you create need to be destroyed.

**Review by bratseth — APPROVED** [review]

**Comment by jonmv** [comment]
Yes, they should be destroyed, but how is this different from using a `DocumentAccess` obtained through `DocumentAccess.createDefault()`?

**Comment by bratseth** [comment]
It's not - so feel free to merge. Just wondering if we couldn't do something better.

**Comment by jonmv** [comment]
Aha, yes, agreed ... :)

File: container-core/src/main/java/com/yahoo/container/core/documentapi/MessageBusDocumentAccessProvider.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        this.documentmanagerConfig = documentmanagerConfig;
+        this.loadTypeConfig = loadTypeConfig;
+    }
+
+    @Override
+    public DocumentAccess get() {
+        synchronized (monitor) {
+            if (access == null) {
+                access = new MessageBusDocumentAccess((MessageBusParams) new MessageBusParams(new LoadTypeSet(loadTypeConfig)).setDocumentmanagerConfig(documentmanagerConfig));
+                if (shutDown)
```
Comment (baldersheim) [inline container-core/src/main/java/com/yahoo/container/core/documentapi/MessageBusDocumentAccessProvider.java:37]: Why creating it at all if shutdown is set. Can you just add it to the outer if as '&& !shutDown'

Comment (jonmv) [reply container-core/src/main/java/com/yahoo/container/core/documentapi/MessageBusDocumentAccessProvider.java:37]: Then you'd return `null`. I thought it better to preserve the existing error message you'd get when trying to use a shut down document access.

Comment (jonmv) [reply container-core/src/main/java/com/yahoo/container/core/documentapi/MessageBusDocumentAccessProvider.java:37]: Alternatively could throw something custom that was clear on this.

Comment (baldersheim) [reply container-core/src/main/java/com/yahoo/container/core/documentapi/MessageBusDocumentAccessProvider.java:37]: Throw instead, no point in constructing mbus access and shut it down just to preserve error message in this case.
This should never happen, but throwing is probably a good idea.

Comment (jonmv) [reply container-core/src/main/java/com/yahoo/container/core/documentapi/MessageBusDocumentAccessProvider.java:37]: Can do. 'twas my first impulse as well. And, yes, this should never happen, unless someone breaks DI.

File: container-core/src/main/java/com/yahoo/container/core/documentapi/MessageBusDocumentAccessProvider.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                    access.shutdown();
+            }
+            return access;
+        }
+    }
+
+    @Override
+    public void deconstruct() {
+        synchronized (monitor) {
+            if ( ! shutDown) {
```
Comment (bjorncs) [inline container-core/src/main/java/com/yahoo/container/core/documentapi/MessageBusDocumentAccessProvider.java:47]: IMHO, throwing an exception for duplicate calls to `deconstruct()` is ok - it should not happen. This is too defensive.

File: container-core/src/main/java/com/yahoo/container/core/documentapi/MessageBusDocumentAccessProvider.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+public class MessageBusDocumentAccessProvider extends AbstractComponent implements Provider<DocumentAccess> {
+
+    private final DocumentmanagerConfig documentmanagerConfig;
+    private final LoadTypeConfig loadTypeConfig;
+    private final Object monitor = new Object();
+    private boolean shutDown = false;
+    private DocumentAccess access = null;
+
+    @Inject
+    public MessageBusDocumentAccessProvider(DocumentmanagerConfig documentmanagerConfig, LoadTypeConfig loadTypeConfig) {
```
Comment (bjorncs) [inline container-core/src/main/java/com/yahoo/container/core/documentapi/MessageBusDocumentAccessProvider.java:27]: Is this config available for all types of container clusters (e.g clustercontroller, metricsproxy, logserver)?

File: config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java [hunk 4]
```diff
@@ -181,6 +181,7 @@ public ContainerCluster(AbstractConfigProducer<?> parent, String subId, String n
         addSimpleComponent(com.yahoo.metrics.simple.jdisc.JdiscMetricsFactory.class.getName(), null, MetricProperties.BUNDLE_SYMBOLIC_NAME);
         addSimpleComponent("com.yahoo.container.jdisc.state.StateMonitor");
         addSimpleComponent("com.yahoo.container.jdisc.ContainerThreadFactory");
+        addSimpleComponent(com.yahoo.container.core.documentapi.MessageBusDocumentAccessProvider.class.getName());
```
Comment (bjorncs) [inline config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java:184]: Should this component be available for non-application-container-clusters?

**Review by bjorncs — COMMENTED** [review]
See comments

**Pull request merged**